Propose GRPCRoute Resource

gnossen opened the pull request:
<!--  Thanks for sending a pull request! Here are some tips for you:

1. If this is your first time contributing to Gateway API, please read our
   developer guide (https://gateway-api.sigs.k8s.io/devguide/)
   and our community page (https://gateway-api.sigs.k8s.io/community/).
2. If this is your first time contributing to a Kubernetes project, please read
   our contributor guidelines:
   https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution
3. Please label this pull request according to what type of issue you are
   addressing, especially if this is a release targeted pull request. For
   reference on required PR/issue labels, read here:
   https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
4. If you want *faster* PR reviews, read how:
   https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it:
   https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind gep

**What this PR does / why we need it**:

Propose a `GRPCRoute` resource.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

howardjohn commented on site-src/geps/gep-XXXX.md:
```diff
@@ -0,0 +1,248 @@
+  rules:
+  - matches:
+      method:
+        grpcService: helloworld.Greeter
```
nit: is `grpc` needed? We know its grpc - we are in a GRPCroute

gnossen replied:
No. Good catch. I had already made this change in the structs, but forgot to reflect it in the example. Fixed.

howardjohn commented on site-src/geps/gep-XXXX.md:
```diff
@@ -0,0 +1,248 @@
+        value: foo
+
+    filters:
+    - type: GRPCRequestHeaderModifierFilter
```
same GRPC prefix comment here and below

gnossen replied:
Fixed.

youngnick commented on site-src/geps/gep-XXXX.md:
```diff
@@ -0,0 +1,248 @@
+# GEP-XXXX: GRPCRoute
```
When you next take a pass on this @gnossen, it's definitely worth logging an "Consider adding a GRPCRoute" issue with the `kind/gep` label, to get a number to put here. Even if we end up not doing this, having this GEP preserved as a record of the discussion is valuable.

brianehlert commented:
Is the intention that this would also incorporate grpc-web in some way? Or the implementation would be strict to grpc.
I am asking since grpc-web is a project technically distinct from grpc with its own nuances.

gnossen commented:
I believe this proposal is now ready for a first round of reviews. @howardjohn @youngnick @robscott

hbagdi commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,813 @@
+precedence. Hence, the GRPCRoute's rule must be evaluated before the HTTPRoute's
+rule.
+
+Accordingly, an xDS server should create something like the following
```
Best to remove specifics of implementation, I don't _think_ we do this anywhere else.

youngnick replied:
I agree we haven't done this very much, but I think the example benefits from showing the eventual LB config in this case - the fact that's an xDS config isn't that important.

gnossen replied:
This section has been removed due to a change in merge semantics.

hbagdi commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,813 @@
+  - bar.com
+  rules:
+  - matches:
+      method:
+        service: helloworld.Greeter
+        method:  SayHello
```
method.method feels off, how about?
```suggestion
      service: helloworld.Greeter
      method:  SayHello
```

gnossen replied:
Do note that there's another field here specifiying the _type_ of the match. Should the `method.type` field become a `methodType` field?

robscott replied:
I'd assumed this only supported an exact match, how would a prefix or regex match work when there are 2 values specified (service and method)?

gnossen replied:
The only two supported types here are "exact" and "regular expression". Here's a full matrix of combinations and their equivalent URI matchers:

## Type Exact

|Service|Method|URI Matcher|
|----------|----------|-----------|
|Specified|Specified|Exact `/${SERVICE}/${METHOD}`|
|Specified|Unspecified|Prefix `/${SERVICE}/`|
|Unspecified|Specified|Suffix `/${METHOD}/` or Regex `/.+/${METHOD}`|
|Unspecified|Unspecified|Prefix `/`|

## Type RegularExpression

|Service|Method|URI Matcher|
|----------|----------|-----------|
|Specified|Specified|Regex `/${SERVICE}/${METHOD}`|
|Specified|Unspecified|Regex `/${SERVICE}/.+`|
|Unspecified|Specified|Regex `/.+/${METHOD}`|
|Unspecified|Unspecified|Prefix `/`|

I'll add this into the proposal to clear this up.

youngnick replied:
Okay, that makes sense to me. Consider my objection resolved, but perhaps something addressing the stutter in the doc to preserve this for later?

robscott replied:
I agree with Nick and Harry here that the stutter does not feel quite right. The explanation here makes me mostly OK with it, but since I think this will be a common critique, it would be good to document the rationale in the GEP.

hbagdi commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,813 @@
+that the user may leave either blank, to be interpreted as an unconditional
+match. Further, this field may either be an exact match or a regex match.
+
+To translate this field to an HTTPRoute path matcher, we propose the following
```
Question: Is all of this translation visible to the end-user or is it implementation guidance?
Is that we expect a controller in the cluster that reconciles GRPCRoute into HTTPRoute?

youngnick replied:
This feels like implementation guidance to me - that the GRPCRoute object should be translated into data-path config equivalent to the object. I am a -1 on actually automatically generating an actual HTTPRoute object.

hbagdi replied:
+1 on your -1

gnossen replied:
This section has been removed due to a change in merge semantics.

hbagdi commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,813 @@
+  path match of `/SERVICE/` with type `PathPrefix`
+- Otherwise, translate the service and the method individually into regular
+  expressions (with an empty field interpreted as `.*`), and translate to an
+  HTTPRoute path matcher of type `RegularExpression` with value `/SERVICE_REGEX/MATCHER_REGEX`
```
There is one missing piece - the protocol. We don't have an explicit HTTP/2 protocol and our current guidance is loose:

https://github.com/kubernetes-sigs/gateway-api/blob/3c8ea05823d3a4215f86b9b3d56316bf2e74a066/apis/v1alpha2/gateway_types.go#L265-L272

One could translate a gRPCRoute to HTTPRoute but if the underlying gateway doesn't support HTTP/2 or has it disabled, none of this would work.

gnossen replied:
I see. I was under the impression that the `HTTP` protocol implied `HTTP2`. I will firm this up.

hbagdi commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,813 @@
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+
+// GRPCRoute provides a way to route gRPC requests. This includes the capability
+// to match requests by hostname, service, method, or header. Filters can be
```
```suggestion
// to match requests by hostname, gRPC-service, gRPC-method, or HTTP header. Filters can be
```

hbagdi commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,813 @@
+// to match requests by hostname, service, method, or header. Filters can be
+// used to specify additional processing steps. Backends specify where matching
+// requests should be routed.
+type GRPCRoute struct {
```
What is the conformance level for this route?

youngnick replied:
Since we only have one conformance profile at the moment, essentially, I'd say its Extended.

gnossen replied:
Added.

hbagdi commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,813 @@
+configuration is received not by an in-line proxy or sidecar proxy but by the client
+itself. Eventually, `GRPCRoute` in the Gateway APIs should support this feature.
+However, to date, there are no HTTP client libraries capable of participating
+in a proxyless service mesh.
```
I'm not very familiar with proxy-less SM. What are the responsibilities of implementation to support this? What changes (if any) are required in this API to support this?

gnossen replied:
Let me start by caveating that proxyless service mesh support is not actually part of *this* proposal.

With that said, I don't believe any changes will be necessary to the `GRPCRoute` resource to enable this. Instead, attachment to a `Mesh` object, referenced through the`parentRefs` field will activate this support. Another proposal for a `Mesh` resource is planned for sometime in the next couple of quarters.

hbagdi commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,813 @@
+	// * Characters in a matching hostname.
+        // * Characters in a matching path.
+	// * Header matches.
+        // * Query param matches.
```
No query-parameter matching is supported yet.
```suggestion
```

gnossen replied:
Hm. This was pulled from the HTTPRoute docstrings. Removed.

hbagdi commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,813 @@
+	// If ties still exist within the Route that has been given precedence,
+	// matching precedence MUST be granted to the first matching rule meeting
+	// the above criteria.
+        //
```
What must a gateway do when a request doesn't match any route?

gnossen replied:
My instinct is that we should do the same thing as HTTPRoute here, but I can't find documentation on what HTTPRoute implementations must do in this case.

hbagdi replied:
Is there a 404 equivalent?

hbagdi commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,813 @@
+// Valid MethodMatchType values are:
+//
+// * "Exact"
+// * "RegularExpression"
```
Conformance levels?

gnossen replied:
This is specified a bit higher up, in the `GRPCMethodMatch` struct.

robscott replied:
Do we need match types here? If so, do they apply to both Service and Method? Should we split those into separate matches?

gnossen replied:
> Do we need match types here?

I think so. See my use cases for regex matches [here](https://github.com/kubernetes-sigs/gateway-api/pull/1004#discussion_r826430565).

> If so, do they apply to both Service and Method?

Yes.

> Should we split those into separate matches?

I don't see the value add in doing this. Again, see [this comment](https://github.com/kubernetes-sigs/gateway-api/pull/1004#discussion_r826430565).

hbagdi commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,813 @@
+	// entries with an equivalent header name MUST be ignored. Due to the
+	// case-insensitivity of header names, "foo" and "Foo" are considered
+	// equivalent.
+	Name GRPCHeaderName `json:"name"`
```
Isn't a gRPC header the same as a HTTP header? My knowledge in this area is weak so please bear with me.

gnossen replied:
This is true, roughly speaking. More details [here](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md).

hbagdi replied:
It may make sense to call this "HeaderName" instead to avoid unnecessary confusion. The "GRPC" prefix doesn't bring any additional value but can cause confusion here.

hbagdi commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,813 @@
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	Filters []GRPCRouteFilter `json:"filters,omitempty"`
```
Since the proposal is specific to gRPC using HTTP/2, what are the downsides of not introducing a new filter type and all the corresponding filters and instead re-using what exists in HTTP?

gnossen replied:
The downside is that then we would not be able to add extra gRPC-specific filters. I imagine most of the items in the "Future Enhancements" section will need to be implemented as new gRPC-specific filters.

pleshakov commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,813 @@
+	// can support POSIX, PCRE, RE2 or any other regular expression dialect.
+	// Please read the implementation's documentation to determine the supported
+	// dialect.
+	PathMatchRegularExpression PathMatchType = "RegularExpression"
```
Would it make sense to also consider matching based on the prefix ? For example, to match based on the package:
```yaml
  rules:
    matches:
    - service: foo. # route traffic that belongs to package foo
      type: Prefix
    backendRefs:
    - name: grpc-backend
      port: 50051
 ```
 
It looks like it is possible to configure that using the RegularExpression match type. However, prefix matching in proxies are more efficient than regex-based matching.

Is it a viable use case?

gnossen replied:
I've considered this. The problem is that allowing prefix matching on the _service_ with a method specified would actually be an HTTP/2 _infix_ matcher. For example, consider the following:

```
  rules:
    matches:
    - service: foo.
      method: bar
      type: Prefix
```

This would match HTTP/2 URIs conforming to this regex:

```
/foo\..*/bar.*
```

It also becomes much more difficult to justify the fact that both the service and the method share a matcher type if you allow prefix match. Even if you had both a `serviceType` and a `methodType`, you still have this infix matcher problem though.

My preference is to keep it simple for the moment and go with what's currently included in the proposal. If folks feel strongly about this though, I'm happy to hear the feedback.

bowei commented:
/assign

youngnick commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,813 @@
+
+### Encapsulated Network Protocols
+
+At the time of writing, the only kind of route officially defined by the Gateway
```
This is a little incorrect, we curently have HTTPRoute, TLSRoute, TCPRoute, and UDPRoute. Could we say that HTTPRoute is the only layer 7 specific route type? Probably.

gnossen replied:
Wasn't actually aware of this. I was going off of the this page, which seems to list only [`HTTPRoute`](https://gateway-api.sigs.k8s.io/v1alpha2/api-types/gatewayclass/). I'll correct this.

youngnick commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,813 @@
+
+- Users of the encapsulated protocol would miss out on significant conventional features from their ecosystem if forced to route at a lower layer.
+- Users of the enapsulated protocol would experience a degraded user experience if forced to route at a lower layer.
+- The encapsulated protocol has a significant user base, particularly in the Kubernetes community.
```
No change here, just wanted to point out that this is a great set of rules that we probably should pull out into a separate doc somewhere.

gnossen replied:
Thanks. Any guidance on placement?

youngnick reviewed (commented):
Not much extra to add on what others have commented.

gnossen commented:
Sorry for the long turnaround time on addressing review comments. I was resolving a discussion on the implementability of this proposal. As a result, I've changed it slightly. We no longer allow hostnames to be shared between `GRPCRoute`s and `HTTPRoute`s. Users will always have the ability to use an `HTTPRoute` if they really must share the hostname and it's a backward-compatible change to allow these in the future if we find strong demand for it and also find a way to make it feasible from a technical perspective. Details are included in the updated proposal.

I've also added two new protocol types: "H2" and "H2C".

gnossen commented:
@youngnick @bowei @pleshakov @hbagdi Friendly ping. Ready for the next round of review.

gnossen commented:
@youngnick @bowei @pleshakov @hbagdi Friendly ping. Waiting on the next round of review.

robscott commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,813 @@
+	//
+	// * HeaderMatches are translated exactly, since the field are
+	//    one-to-one
+	// * If `MethodMatch.type` is `Exact` and neither the
+	//    `MethodMatch.service` nor `MethodMatch.method` fields are present,
+	//    translate to a `PathMatch` of `{"type": "PathPrefix", value: "/"}`
```
Would exact match on "/" be a better translation here?

gnossen replied:
Removed this section since we're no longer proposing merge between `GRPCRoute` and `HTTPRoute` (but my last `git push` failed -- whoopsie)

Regardless, this semantic still applies. I think an unspecified method and unspecified match should match _all_ gRPC traffic. Do you think there's an issue with this?

An exact match on `/` is not even a valid gRPC+protobuf URI, so I don't think this makes much sense.

robscott commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,813 @@
+	// * If `MethodMatch.type` is `Exact` and the `MethodMatch.service`
+	//    field is present but not the `MethodMatch.method` field, translate to a
+	//    `PathMatch` of `{"type": "PathPrefix", value: "/" + service}`
+	// - Otherwise, translate the service and the method individually into
+	//    regular expressions (with an absent field interpreted as `.*`) and
+	//    translate to a `PathMatch` of
+	//    `{"type": "RegularExpression", value: "/": serviceRegex + "/" + methodRegex}`
```
To clarify, this would essentially translate an empty Method match with a `foo` Service to `type: PathPrefix, value: /foo`, and a regex would only be required if the match specified a path but not a Service? Not very familiar with gRPC, but is that last option a very common one? Would it be valid to just not support that?

gnossen replied:
> To clarify, this would essentially translate an empty Method match with a foo Service to type: PathPrefix, value: /foo

Correct.

> a regex would only be required if the match specified a path but not a Service

You can imagine many different use cases for regexes. Many gRPC APIs are resource oriented. You may want to route only methods that start with `Get`. In this case, you'd want to do something like:

```
method:
 type: RegularExpression
 service: foo.org
 method: Get.+
```

Or you might do something similar with services. Say you want to route both v1 and v2 of a service. What's nice here is that exact matches are a subset of regular expressions, so the fact that there's only one type for two values here doesn't actually get in the way of expressing the routing logic you want.

> Would it be valid to just not support that?

It would be valid, but it would also be limiting. This is why I have listed "Custom" support for the RegularExpression matcher type.

robscott commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,813 @@
+	// Support: Core
+	//
+	// +optional
+	RequestHeaderModifier *GRPCRequestHeaderFilter `json:"requestHeaderModifier,omitempty"`
```
Can we reuse the same type we have for HTTP header modifier here?

gnossen replied:
Sure. That seems reasonable. I wasn't sure if there was some rule against sharing filters between different route types.

youngnick replied:
I think the expectation is that unsupported entries in this list should result in admission failure for the resource, just like it does for HTTPRoute, so reusing the same type spec should be fine.

gnossen replied:
Switched `GRPCRequestHeaderFilter` to `HTTPRequestHeaderFilter` and `GRPCRequestMirrorFilter` to `HTTPRequestMirrorFilter`.

robscott reviewed (commented):
Thanks for your work on this @gnossen!

youngnick commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,768 @@
+If the listener supports neither `H2` nor `H2C`, then the implementation MUST
+raise an `Accepted` Condition with a status of `False` in the corresponding
+RouteParentStatus.
+
```
I know we don't always ask for this, but because this is the first time we're adding a new Route resource, I'd like to see something about the graduation plan.

Based on what we have in #1039, I'd suggest something like:
- gRPC resource implemented by a couple of implementations
- Some conformance tests in place
- Users are using the API surface
- API review done by upstream reviewers.

Graduation to GA should be similar but more stringent (exhaustive conformance tests, implemented by lots o implementations, etc).

youngnick reviewed (commented):
This is looking great, just a couple of outstanding questions from me. Sorry about the review delay.

gnossen commented:
@youngnick @bowei @pleshakov @hbagdi Friendly ping. Ready for the next round of review.

youngnick reviewed (commented):
This LGTM with one small request to record the discussion about stuttering so we don't have to rehash it at some point.

bowei commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,781 @@
+
+## Introduction
+
+While it would be possible to support gRPC via CRDs, in the long run, this would
```
To be clear, the whole API is done with CRDs. Do you mean custom, out-of-tree resources?

gnossen replied:
Yes. Clarified.

bowei commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,781 @@
+- Many features supported by HTTP/2 but not by gRPC, such as
+  - Query parameters
+  - Methods besides `POST`
+  - CORS
```
CORS headers (might want to add a link https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS),

gnossen replied:
Done.

bowei commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,781 @@
+#### Proxyless Service Mesh
+
+The gRPC library supports proxyless service mesh, a system by which routing
+configuration is received not by an in-line proxy or sidecar proxy but by the client
```
nit: configuration and routing is implemented by the client stub itself (receiving is just one part of the story)

gnossen replied:
Clarified.

bowei reviewed (commented):
LGTM 

We can take the specific API review comments to the actual API PR

youngnick commented:
/hold for maintainer LGTMs

@hbagdi @howardjohn, do you have any further comments? We'll give one more week starting from the time this is posted.

I'll LGTM once my comments are addressed, @bowei can do the same, and then once we have one more maintainer LGTM, we are good to go.

bowei commented:
/lgtm

gnossen commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,813 @@
+
+#### HTTP/2 Cleartext
+
+gRPC allows HTTP/2 cleartext communication (H2C). This is conventionally deployed for
```
This _is_ in scope. Note the two new added `ProtocolType` options.

gnossen commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,813 @@
+
+#### Content-Based Routing
+
+While not included in the scope of this initial GEP, a common use case cited for
```
I'm not sure I follow? This sentence is calling out a commonly requested use case for `GRPCRoute` as a justification for its existence while simultaneously calling that use case out of scope for this GEP.

gnossen commented:
Sorry for the delay on responding to these last few review comments.

@bowei @youngnick I believe I've addressed everything you've brought up.

gnossen commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,694 @@
+        UDPProtocolType ProtocolType = "UDP"
++
++       // Accepts HTTP/2 sessions over TLS.
++       H2ProtocolType ProtocolType = "H2"
```
Because `HTTPS` doesn't necessarily imply support for HTTP/2. You could have an `HTTPS` implementation that only supports HTTP/1.1. An implementation supporting `H2` would guarantee that HTTP/2 over TLS with no upgrade from HTTP/1 works.

howardjohn replied:
It seems a bit inconsistent... so we now have:

* `HTTP`: h1, optionally h2c
* `HTTPS`: h1 or h2 (over TLS)
* `H2C`: only h2c
* `HTTP2`: only h2 (over TLS)

I can't imagine any user could see those 4 options and understand what they mean without a close inspection of the docs

Also...
> You could have an HTTPS implementation that only supports HTTP/1.1

Doesn't the spec not allow that?

```
	// Accepts HTTP/1.1 or HTTP/2 sessions over TLS.
	HTTPSProtocolType ProtocolType = "HTTPS"
```

> An implementation supporting H2 would guarantee that HTTP/2 over TLS with no upgrade from HTTP/1 works.

It guarantees HTTP/2 works, but does it guarantee HTTP/1 does *not* work?

howardjohn replied:
How does it work when I have `HTTPS` and `H2` (or `HTTP` and `H2C`) on the same port?

If we say its a conflict, then that means that if I want to do h1 and h2 on the same port, I need to name the port `HTTP`, not `H2C`. So we will need to tell users "If you want to use HTTP2 over clear text, do NOT set protocol: h2c".

If its not a conflict and they can merge, then that would imply we can accept HTTP/1.1 over the H2C/HTTP2 protocols which seems unexpected?

gnossen replied:
> Doesn't the spec not allow that?

The word "or" is doing a lot of work here. If we should be interpreting this to mean "Accepts HTTP/1.1 or (HTTP/1.1. and HTTP/2)", then I suppose so.

> How does it work when I have HTTPS and H2 (or HTTP and H2C) on the same port?

I would expect `HTTPS` and `H2` on the same port to mean that the port would accept any of the following over TLS:

- HTTP/1.1
- HTTP/2 via upgrade from HTTP/1.1
- HTTP/2 directly (i.e. without upgrade from HTTP/1.1)

Whereas `H2` on its own would guarantee only:

- HTTP/2 directly (i.e. without upgrade from HTTP/1.1)

While `HTTPS` would guarantee:

- HTTP/1.1
- HTTP/2 via upgrade from HTTP/1.1

> If its not a conflict and they can merge, then that would imply we can accept HTTP/1.1 over the H2C/HTTP2 protocols which seems unexpected?

I don't think this is true. A listener with both `HTTPS` and `H2` that accepts HTTP/1.1 would be doing so under the capabilities of `HTTPS`, not of `H2`. This seems similar to what you'd expect for a port that accepts both TCP and HTTP (or is this not allowed?)


More generally, I need to solve two problems here:

- Providing an indication that an implementation definitely supports HTTP/2 (this may be solved already, as you pointed out, depending on how you read the current spec for HTTPS)
- Providing an indication that an implementation definitely accepts HTTP/2 without first upgrading from HTTP/1. Not all servers do this.

If you can propose an alternative that accomplishes these two goals, I'd be more than happy to adopt it.

howardjohn replied:
When you say upgrade are you referring to https://en.wikipedia.org/wiki/HTTP/1.1_Upgrade_header#Use_with_HTTP/2 or something else? Since "The HTTP Upgrade mechanism is used only for cleartext HTTP2 (h2c)"

howardjohn replied:
Given both the HTTP and HTTPS protocol, in the current spec, allow HTTP2, which don't we just say that GRPC can only bind to HTTP/HTTPS, and the GRPCRoutes only accept HTTP/2 requests.

I don't see why we need 'Providing an indication that an implementation definitely supports HTTP/2'? if the implementation accepts GRPCRoutes then they must accept HTTP/2 (we can declare that in the spec if needed)

gnossen replied:
> When you say upgrade are you referring to https://en.wikipedia.org/wiki/HTTP/1.1_Upgrade_header#Use_with_HTTP/2

Yes.

> if the implementation accepts GRPCRoutes then they must accept HTTP/2 (we can declare that in the spec if needed)

I was under the impression that this sort of requirement was expected to be expressed through the `ProtocolType`. I have no problem with this just being part of the spec for GRPCRoute.

I have switched to making this requirement in the spec for GRPCRoute and have removed the `H2` and `H2C` ProtocolTypes entirely.

hbagdi commented on site-src/geps/gep-1016.md:
````diff
@@ -0,0 +1,685 @@
+	// ```
+	// matches:
+	// - method:
+	//     method: foo.bar
````
nit: `foo.bar` a valid gRPC method name?

gnossen replied:
Ah, good catch. Should be `service` here, not `method`. Fixed.

hbagdi commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,685 @@
+	//
+	// Specifying a core filter multiple times has unspecified or custom
+	// conformance.
+	// Support: Core
```
This has come before so I want to clarify.
Since gRPC filters are part of a bigger feature, gRPCRoute, they shouldn't have a conformance level that is higher than the parent.

Should this be `extended`?

gnossen replied:
I think I need some guidance here.

> Since gRPC filters are part of a bigger feature, gRPCRoute, they shouldn't have a conformance level that is higher than the parent.

I absolutely agree with this. In the places where I've written "Core" in this proposal, I've meant "Core to GRPCRoute."  Likewise, "Custom" here means "Custom to GRPCRoute."

If I write the support levels in absolute terms, then I think that would flatten things out and everything would become "Support: extended." This loses much of the detail that I've written in about support levels between different GRPCRoute implementations.

Thoughts on how to resolve this?

hbagdi replied:
> I absolutely agree with this. In the places where I've written "Core" in this proposal, I've meant "Core to GRPCRoute." Likewise, "Custom" here means "Custom to GRPCRoute."

That is how I read this as well but we run into issues such as #655.

> If I write the support levels in absolute terms, then I think that would flatten things out and everything would become "Support: extended." This loses much of the detail that I've written in about support levels between different GRPCRoute implementations.

Extended features must be portable _if_ they are supported. Is that not enough for defining semantics here?

cc @robscott @youngnick

youngnick replied:
I don't think we've explicitly discussed this before, but I've always conceived of the field-level support working like @gnossen described - this is because we clearly _have_ to allow implementations not to support some objects, because Layer 7 ingress controllers may not be able to do Layer 4 resources (like TCPRoute and UDPRoute), and Layer 4 load balancers may not be able to do higher-layer resources (like HTTPRoute or TLSRoute).

So, the "Support: Core" markings on fields have to be on a per-object basis.

But like I said, this is a subtle point that I don't think we've explicitly made before.

robscott replied:
I agree with the interpretation here. We do cover this in the docs, but it may not be prominent enough (https://gateway-api.sigs.k8s.io/concepts/guidelines/#overlapping-support-levels).

hbagdi commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,685 @@
+// GRPCPathMatch describes how to select a gRPC route by matching the gRPC
+// request service and/or method..
+type GRPCMethodMatch struct {
+	// Type specifies how to match against the service and/or method.
+	// Support: Core (Exact with method specified)
+	//
+	// Support Custom (Exact with no method specified)
+	//
+	// Support: Custom (RegularExpression)
+	//
+	// +optional
+	// +kubebuilder:default=Exact
+	Type *GRPCMethodMatchType `json:"type,omitempty"`
+
+
+	// Value of the service to match against. If left empty or omitted, will
+	// match all services.
+	// +optional
+	// +kubebuilder:default=""
+	// +kubebuilder:validation:MaxLength=1024
+	Service *string `json:"value,omitempty"`
+
+	// Value of the method to match against. If left empty or omitted, will
+	// match all services.
+	// +optional
+	// +kubebuilder:default=""
+	// +kubebuilder:validation:MaxLength=1024
+	Method *string `json:"value,omitempty"`
```
I find this pretty confusing and we should do a bit of optimization for the easy path.

Based on the defaults, the values are:
```
type: Exact
method: ""
service: ""
```

No matter what we document, this - 'exact match on empty strings' makes little sense to me and I _feel_ it will make little sense to anyone.
If the type is Exact, then the values should contain something.
My experience is that using empty values to mean everything doesn't work well, especially in k8s.

Some options we can explore:
1. Can the code require at least one of method or service?
2. Can we ask the user to provide an explicit `*` as the value of the `method` or `service` to mean that they would like to match any method or service respectively? IMHO, that makes the configuration much more self-documenting.
3. If type=Exact, the code should be exacting. Splitting `type` into two types makes the configuration verbose but it also makes it a bit more machine-readable. Let's say the type=regex and then one of them is a regex while the other one is not, how is the code supposed to figure it out?
4. Exact with the method specified is core and no method specified is custom seems arbitrary and confusing. Can we classify both at the same level?

I should note that the above is done with the right technical reasons, I don't doubt. It doesn't feel it is intuitive enough.

Thoughts?

gnossen replied:
> Based on the defaults, the values are:

This is true. Those are the default values. But people generally don't see the default values, they see the default _behavior_. And the default _behavior_ is to match all traffic.

> My experience is that using empty values to mean everything doesn't work well, especially in k8s.

Can you provide some concrete examples?

> Can the code require at least one of method or service?

Considering that supplying no match at all already matches all traffic, I'm totally open to this option. Going ahead and making this change. Anyone objecting, please jump in.

> Can we ask the user to provide an explicit * as the value of the method or service to mean that they would like to match any method or service respectively? IMHO, that makes the configuration much more self-documenting.

I'm not the biggest fan of this option. As far as I know, no equivalent syntax exists for HTTPRoute. This could lead to confusion.

> If type=Exact, the code should be exacting. Splitting type into two types makes the configuration verbose but it also makes it a bit more machine-readable.

I don't quite see how this solves the problem you introduced at the beginning of this comment? And as you point out, this would not be as nice a UX.

> Let's say the type=regex and then one of them is a regex while the other one is not, how is the code supposed to figure it out?

Discussed already [here](https://github.com/kubernetes-sigs/gateway-api/pull/1004#discussion_r826430565).

Specifically, this:

> What's nice here is that exact matches are a subset of regular expressions, so the fact that there's only one type for two values here doesn't actually get in the way of expressing the routing logic you want. 

In practice, this shouldn't be a limitation.

> Exact with the method specified is core and no method specified is custom seems arbitrary and confusing. Can we classify both at the same level?

This actually was mistaken. The combination that presents trouble for implementations is `type: EXACT` with service unspecified and method specified. I've corrected this in the proposal.

I expect that many implementations will add support by translating to HTTPRoutes. In [this chart](https://github.com/kubernetes-sigs/gateway-api/pull/1004#discussion_r826425912) from earlier, you see that this combination is equivalent to either a suffix matcher or a regex matcher of the form `/.+/${METHOD}`. Suffix matchers don't exist for HTTPRoute and Regex matchers have "Custom" support.  This case is called out to have the same support level as an HTTPRoute regex matcher to allow implementations that support HTTPRoute but not regex to also support GRPCRoute at some level.

If you have an alternative to address this problem, I'm open to suggestions.

hbagdi replied:
> This is true. Those are the default values. But people generally don't see the default values, they see the default behavior. And the default behavior is to match all traffic.

In that case, the default values _should_ communicate the default behavior well.

> Can you provide some concrete examples?

Empty label selector behavior comes to mind.


> Considering that supplying no match at all already matches all traffic, I'm totally open to this option. Going ahead and making this change. Anyone objecting, please jump in.

Great. Could you create a tracking issue for this? This will require updates to validating webhook; the last time I looked, controller-runtime didn't allow for "at least one" JSON-schema validations. 


> > Can we ask the user to provide an explicit * as the value of the method or service to mean that they would like to match any method or service respectively? IMHO, that makes the configuration much more self-documenting.

> I'm not the biggest fan of this option. As far as I know, no equivalent syntax exists for HTTPRoute.

In HTTPRoute, the default value for path is `/` with the type `PathPrefix`.

gnossen replied:
> Great. Could you create a tracking issue for this? This will require updates to validating webhook; the last time I looked, controller-runtime didn't allow for "at least one" JSON-schema validations.

[Created.](https://github.com/kubernetes-sigs/gateway-api/issues/1098)

gnossen commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,813 @@
+
+#### Content-Based Routing
+
+While not included in the scope of this initial GEP, a common use case cited for
+routing gRPC is payload-aware routing. That is, routing rules which determine a
+backend based on the contents of the protocol buffer payload.
```
Yes. Note added about this.

youngnick commented:
/lgtm

but will leave this for @hbagdi to come back to for approval and his lgtm

robscott commented:
/ok-to-test

hbagdi commented on site-src/geps/gep-1016.md:
```diff
@@ -0,0 +1,704 @@
+// At least one of Service and Method must be a non-empty string.
+type GRPCMethodMatch struct {
+	// Type specifies how to match against the service and/or method.
+	// Support: Core (Exact)
```
```suggestion
	// Support: Core (Exact with service and method specified)
```

robscott commented:
Thanks for the work on this @gnossen! This is a well written and massive GEP. I'm deferring to @hbagdi for final LGTM.

/approve

hbagdi commented:
/lgtm

robscott commented:
Looks like we're good to merge this in, thanks for all the work on this @gnossen!

/hold cancel
